File: Ast.hs
module Ast where

type Name = String
type Parameter = String
type Binding = (String, Expression)
type Length = Integer

data Program = Program [Variable] [Callable]

data Module = Module [FilePath] [Variable] [Callable]

data Element = ImportElement FilePath
             | VariableElement Variable
             | CallableElement Callable

data Variable = IntegerVariable       { variableName :: Name 
                                      , integerValue :: Integer }
              | ArrayVariable         { variableName :: Name 
                                      , arrayValue :: [Integer] }
              | StringVariable        { variableName :: Name 
                                      , stringValue :: String }
              | UninitializedVariable { variableName :: Name 
                                      , uninitializedLength :: Length }

variableLength :: Variable -> Integer
variableLength (IntegerVariable _ _)         = 1
variableLength (ArrayVariable _ xs)          = toInteger (length xs)
variableLength (StringVariable _ str)        = toInteger $ (length str) + 1
variableLength (UninitializedVariable _ len) = len
                                      
data Callable = Function { callableName     :: Name
                         , callableParams   :: [Parameter]
                         , functionBindings :: [Binding]
                         , functionBody     :: Expression }
              | Macro    { callableName   :: Name
                         , callableParams :: [Parameter]
                         , macroBody      :: [Instruction] }

data Expression = IntegerLiteral Integer
                | StringLiteral String
                | Symbol Name
                | CallEx Name [Expression]
                | If Expression Expression Expression
                | Do [Expression]

data Instruction = Push Integer
                 | Dup
                 | Copy Integer
                 | Swap
                 | Pop
                 | Slide Integer
                 | Add
                 | Sub
                 | Mul
                 | Div
                 | Mod
                 | Store
                 | Load
                 | Label String
                 | Call String
                 | Jump String
                 | Jz String
                 | Jn String
                 | Ret
                 | End
                 | Pc
                 | Pn
                 | Rc
                 | Rn

fromElements :: [Element] -> Module
fromElements elements = Module imports variables callables 
    where imports = [i | ImportElement i <- elements]
          variables = [v | VariableElement v <- elements]
          callables = [c | CallableElement c <- elements]
        
fromModules :: [Module] -> Program
fromModules modules = Program variables callables
    where variables = concat $ [v | Module _ v _ <- modules]
          callables = concat $ [c | Module _ _ c <- modules]
          

instance Show Program where
    show (Program variables callables) = 
        (unlines $ map show variables) ++ "\n" ++
        (unlines $ map show callables)
        
instance Show Variable where
    
    show (IntegerVariable name i) = 
        "var " ++ name ++ " = " ++ (show i)
        
    show (ArrayVariable name is) = 
        "var " ++ name ++ " = (" ++ (unwords $ map show is) ++ ")"
        
    show (StringVariable name str) = 
        "var " ++ name ++ " = " ++ (show str)
        
    show (UninitializedVariable name len) = 
        "var " ++ name ++ "[" ++ (show len) ++ "]"
        
instance Show Callable where

    show (Function name params bindings body) = 
        "def " ++ name ++ " (" ++ (unwords params) ++ ") = " ++
        (showBindings bindings) ++
        (show body) ++ "\n"
      where showBindings [] = ""
            showBindings bs = "let " ++ (unlines $ map showBinding bs) ++ " in "
            showBinding (name, value) = name ++ " = " ++ (show value)
            
    show (Macro name params instructions) = 
        "asm " ++ name ++ " (" ++ (unwords params)  ++ ") = (\n" ++
        (unlines $ map show instructions) ++ ")\n"
   
instance Show Expression where

    show (IntegerLiteral i) = (show i)
    show (StringLiteral str) = (show str)
    show (Symbol name) = name
    
    show (CallEx name args) = 
        "(" ++ name ++ " " ++ (unwords $ map show args) ++ ")"

    show (If c t f) = 
        "(if " ++ (show c) ++ "\n" ++ (show t) ++ "\n" ++ (show f) ++ ")\n"
        
    show (Do exps) = 
        "(do " ++ (unlines $ map show exps) ++ ")\n"

instance Show Instruction where
    show (Push i)    = "push " ++ (show i)
    show (Dup)       = "dup"
    show (Copy i)    = "copy " ++ (show i)
    show (Swap)      = "swap"
    show (Pop)       = "pop"
    show (Slide i)   = "slide " ++ (show i)
    show (Add)       = "add"
    show (Sub)       = "sub"
    show (Mul)       = "mul"
    show (Div)       = "div"
    show (Mod)       = "mod"
    show (Store)     = "store"
    show (Load)      = "load"
    show (Label lbl) = "label " ++ lbl
    show (Call lbl)  = "call " ++ lbl
    show (Jump lbl)  = "jump " ++ lbl
    show (Jz lbl)    = "jz " ++ lbl
    show (Jn lbl)    = "jn " ++ lbl
    show (Ret)       = "ret"
    show (End)       = "end"
    show (Pc)        = "pc"
    show (Pn)        = "pn"
    show (Rc)        = "rc"
    show (Rn)        = "rn"

File: ast.py
from escape import escape

class Program:

    def __init__(self, variables, functions):
        self.variables = variables
        self.functions = functions
        
    def __repr__(self):
        return ('\n'.join(map(repr, self.variables)) + '\n\n' +
                '\n'.join(map(repr, self.functions)))
        
class Module:

    def __init__(self, imports, variables, functions):
        self.imports = imports
        self.variables = variables
        self.functions = functions
        
    def __repr__(self):
        return ('\n'.join(map(repr, self.imports)) + '\n\n' +
                '\n'.join(map(repr, self.variables)) + '\n\n' +
                '\n'.join(map(repr, self.functions)))
        
class Import:

    def __init__(self, token, file):
        self.token = token
        self.file = file
        
    def __repr__(self):
        return 'import "' + self.file + '"'
        
class Variable:
    pass
    
class IntegerVariable(Variable):

    def __init__(self, token, name, value):
        self.token = token
        self.name = name
        self.value = value
        
    def __repr__(self):
        return 'var ' + self.name + ' = ' + str(self.value)
        
class ArrayVariable(Variable):
    
    def __init__(self, token, name, values):
        self.token = token
        self.name = name
        self.values = values
        
    def __repr__(self):
        return 'var ' + self.name + ' = (' + ' '.join(map(str, self.values)) + ')'
        
class StringVariable(Variable):

    def __init__(self, token, name, string):
        self.token = token
        self.name = name
        self.string = string
        
    def __repr__(self):
        return 'var ' + self.name + ' = "' + escape(self.string) + '"'
        
class UninitializedVariable(Variable):

    def __init__(self, token, name, size):
        self.token = token
        self.name = name
        self.size = size

    def __repr__(self):
        return 'var ' + self.name + '(' + str(self.size) + ')'
        
class Function:
    
    def __init__(self, token, inline, name, parameters, bindings, body):
        self.token = token
        self.inline = inline
        self.name = name
        self.parameters = parameters
        self.bindings = bindings 
        self.body = body
        
    def makeRepr(self, strFuncType, strBody):
        
        strFuncType = strFuncType + " "
        strBody = strBody + "\n"
        
        if self.inline:
            strInline = "inline "
        else:
            strInline = ""
            
        strSignature = self.name + " (" + ' '.join(self.parameters) + ") = "
        
        if self.bindings == []:
            strLetForm = ''
        else:
            strLetForm = ("let\n" +
                          '\n'.join([name + ' = ' + str(value) 
                                     for (name, value) in self.bindings]) + '\n' +
                          "in ")
        
        return strFuncType + strInline + strSignature + strLetForm + strBody
    
class HplFunction(Function):
    
    def __repr__(self):
        return self.makeRepr("def", repr(self.body))
        
class AsmFunction(Function):

    def __repr__(self):
        strBody = "(\n" + '\n'.join([str(ins) for ins in self.body]) + "\n)"
        return self.makeRepr("asm", strBody)
        
class Expression:
    pass
    
class IntegerLiteralEx(Expression):
    
    def __init__(self, token, value):
        self.token = token
        self.value = value
        
    def __repr__(self):
        return str(self.value)
        
class StringLiteralEx(Expression):

    def __init__(self, token, string):
        self.token = token
        self.string = string
        
    def __repr__(self):
        return '"' + escape(self.string) + '"'
        
class SymbolEx(Expression):

    def __init__(self, token, name):
        self.token = token
        self.name = name
        
    def __repr__(self):
        return self.name
        
class IfEx(Expression):

    def __init__(self, token, condition, trueValue, falseValue):
        self.token = token
        self.condition = condition
        self.trueValue = trueValue
        self.falseValue = falseValue
        
    def __repr__(self):
        return ('(if ' + repr(self.condition) + '\n' +
                         repr(self.trueValue) + '\n' + 
                         repr(self.falseValue) + ')')
        
class DoEx(Expression):

    def __init__(self, token, expressions):
        self.token = token
        self.expressions = expressions

    def __repr__(self):
        return ('(do ' + '\n'.join(map(repr, self.expressions)) + ')')
        
class CallEx(Expression):
    
    def __init__(self, token, name, arguments):
        self.token = token
        self.name = name
        self.arguments = arguments

    def __repr__(self):
        return '(' + self.name + ' ' + ' '.join(map(repr, self.arguments)) + ')'
        
class Instruction:
    
    def __init__(self, token, command, operand=None):
        self.token = token
        self.command = command
        self.operand = operand

    def __repr__(self):
        if self.operand == None:
            return self.command
        else:
            return self.command + ' ' + str(self.operand)